src/main/java/com/microsoft/sqlserver/jdbc/SQLServerStatement.java

```diff
@@ -631,7 +631,9 @@ void closeInternal() {
         // Regardless what happens when cleaning up,
         // the statement is considered closed.
         assert !bIsClosed;
-
+        
+        connection.removeOpenStatement(this);
```

Comment:
This should be done right at the end of `Statement.close()` just before logging to avoid losing references of statements if they were not closed properly (failure possible in `discardLastExecutionResults`) due to any reason in a pooled connection scenario. It could lead into losing references/memory leakage of unclosed statements/resultsets.

Reply:
Agreed.